include/deal.II/lac/constraint_matrix.templates.h

```diff
@@ -2319,10 +2330,42 @@ ConstraintMatrix::distribute_local_to_global (
                                                    local_matrix);
 
           if (val != number ())
-            global_vector(row) += static_cast<typename VectorType::value_type>(val);
+            {
+              rows[local_row_n] = row;
+              vector_values[local_row_n] = val;
+              ++local_row_n;
+            }
         }
     }
 
+  const size_type n_local_rows = local_row_n;
+  // Attempt to use the faster code path for PETSc if we are dealing with
+  // PetscScalars.
+#ifdef DEAL_II_WITH_PETSC
+  PETScWrappers::VectorBase *petsc_global_vector
+    = dynamic_cast<PETScWrappers::VectorBase *>(&global_vector);
+  if (petsc_global_vector != 0 && types_are_equal<PetscScalar, number>::value)
+    {
+      rows.resize(n_local_rows);
+      vector_values.resize(n_local_rows);
+      petsc_global_vector->add(rows, *reinterpret_cast<std::vector<PetscScalar> *>(&vector_values));
+    }
+  else
+    {
+      for (size_type row_n=0; row_n<n_local_rows; ++row_n)
+        {
+          global_vector(rows[row_n]) +=
+            static_cast<typename VectorType::value_type>(vector_values[row_n]);
```

Comment:
Out of curiosity: Don't the other vector classes have an `add` function as well?